rust/windows-client/src-tauri/src/client/device_id.rs

```diff
@@ -0,0 +1,44 @@
+use ring::digest;
+use tokio::fs;
+
+/// Get the hashed device ID, generating it if it's not already on disk.
+/// Per <https://github.com/firezone/firezone/issues/2697> and <https://github.com/firezone/firezone/issues/2711>,
+/// clients must generate their own random IDs and persist them to disk, to handle situations like VMs where a hardware ID is not unique or not available.
+///
+/// Returns: The hexadecimal SHA256 of the UUID, suitable for sending directly to `connlib_client_shared::Session::connect`.
+/// Errors: If the disk is unwritable when initially generating the ID, or unreadable when reading it back, or if the file is not valid JSON or doesn't match the expected schema
+pub(crate) async fn hashed_device_id(
+    app_local_data_dir: &crate::client::AppLocalDataDir,
+) -> anyhow::Result<String> {
+    let dir = app_local_data_dir.0.join("config");
+    let path = dir.join("device_id.json");
+
+    // Try to read it back from disk
+    if let Ok(s) = fs::read_to_string(&path).await {
+        let j: DeviceIdJson = serde_json::from_str(&s)?;
```

Comment:
I think if it's invalid here we should just generate a new one.

Reply:
Done in 98a80be8